src/main/java/net/minestom/server/listener/preplay/HandshakeListener.java

```diff
@@ -48,74 +51,40 @@ public static void listener(@NotNull ClientHandshakePacket packet, @NotNull Play
-                            final String rawPropertyJson = split[3];
-                            final JsonArray propertyJson = JsonParser.parseString(rawPropertyJson).getAsJsonArray();
-                            for (JsonElement element : propertyJson) {
-                                final JsonObject jsonObject = element.getAsJsonObject();
-                                final JsonElement name = jsonObject.get("name");
-                                final JsonElement value = jsonObject.get("value");
-                                final JsonElement signature = jsonObject.get("signature");
-                                if (name == null || value == null) continue;
-
-                                final String nameString = name.getAsString();
-                                final String valueString = value.getAsString();
-                                final String signatureString = signature == null ? null : signature.getAsString();
-
-                                if (BungeeCordProxy.isBungeeGuardEnabled() && nameString.equals("bungeeguard-token")) {
-                                    if (foundBungeeGuardToken || !BungeeCordProxy.isValidBungeeGuardToken(valueString)) {
-                                        bungeeDisconnect(socketConnection);
-                                        return;
-                                    }
-
-                                    foundBungeeGuardToken = true;
-                                }
-
-                                properties.add(new GameProfile.Property(nameString, valueString, signatureString));
-                            }
+                if (!MinecraftServer.getConnectionRule().isAllowed(ClientHandshakePacket.Intent.LOGIN)) {
+                    // Direct connections not allowed
+                    disconnect(connection, DIRECT_CONNECTIONS_NOT_ALLOWED);
+                }
 
-                            if (BungeeCordProxy.isBungeeGuardEnabled() && !foundBungeeGuardToken) {
-                                bungeeDisconnect(socketConnection);
-                                return;
-                            }
-                        }
+                // Bungee support (IP forwarding)
+                String bungeeAddress = bungeeConnect(address, connection);
 
-                        final GameProfile gameProfile = new GameProfile(playerUuid, "test", properties);
-                        socketConnection.UNSAFE_setProfile(gameProfile);
-                    } else {
-                        bungeeDisconnect(socketConnection);
-                        return;
-                    }
+                if (bungeeAddress == null) {
+                    return;
                 }
+
+                address = bungeeAddress;
             }
             case TRANSFER -> {
-                throw new UnsupportedOperationException("Transfer intent is not supported in HandshakeListener");
+                connection.setConnectionState(ConnectionState.LOGIN);
```

Comment:
The login event should have an isTransfer field which would allow people to handle disallowing transfers if they choose.